### PR TITLE
[compiler-v2] Revise copy creation in ability processor

### LINKED ISSUE #13488: [Bug][compiler-v2] MOVELOC_EXISTS_BORROW_ERROR raised when generating non-opt code for copy_loc_borrowed_indirect
# 🐛 Bug

When compiling the following code using V2:

```
module 0x8675309::Tester {
    fun t() {
        let x = 0;
        let y = 0;
        let r1 = foo(&x, &y);
        let r2 = foo(&x, &y);
        x + copy x;
        y + copy y;
        r1;
        r2;
    }

    fun foo(_r: &u64, r2: &u64): &u64 {
        r2
    }
}
```

The no-opt.exp file will contain a bytecode verification error while the exp file does not:

```
Diagnostics:
bug: BYTECODE VERIFICATION FAILED
  ┌─ tests/reference-safety/v1-borrow-tests/copy_loc_borrowed_indirect.move:7:9
  │
7 │         x + copy x;
  │         ^^^^^^^^^^ ICE failed bytecode verifier: VMError {
    major_status: MOVELOC_EXISTS_BORROW_ERROR,
    sub_status: None,
    message: None,
    exec_state: None,
    location: Module(
        ModuleId {
            address: 0000000000000000000000000000000000000000000000000000000008675309,
            name: Identifier(
                "Tester",
            ),
        },
    ),
    indices: [
        (
            FunctionDefinition,
            1,
        ),
    ],
    offsets: [
        (
            FunctionDefinitionIndex(1),
            12,
        ),
    ],
}
```

### LINKED ISSUE #14227: [Bug] MOVELOC_UNAVAILABLE_ERROR is raised when the enum type is not copyable
# 🐛 Bug

When compiling the following code:

```
module 0x815::m {

    enum CommonFields has drop {
    Foo{x: u64, y: u8},
    Bar{x: u64, y: u8, z: u32}
    }

    enum CommonFieldsVector has drop {
    Foo{x: vector<u8>},
    Bar{x: vector<u8>, y: vector<CommonFields>}
    }

    fun test_enum_vector() {
        let _common_fields = CommonFields::Bar {
            x: 30,
            y: 40,
            z: 50
        };
        let _common_vector_2 = CommonFieldsVector::Bar {
            x: vector[2],
            y: vector[_common_fields]
        };
        let _common_vector_3 = CommonFieldsVector::Bar {
            x: vector[2],
            y: vector[_common_fields]
        };

    }
}

```

Instead of complaining `_common_fields` cannot be implicitly copied, the bytecode verification error `MOVELOC_UNAVAILABLE_ERROR` is raised:

```
bug: bytecode verification failed with unexpected status code `MOVELOC_UNAVAILABLE_ERROR`. This is a compiler bug, consider reporting it.
   ┌─ tests/bytecode-generator/enum.move:21:32
   │
21 │           let _common_vector_2 = CommonFieldsVector::Bar {
   │ ╭────────────────────────────────^
22 │ │             x: vector[2],
23 │ │             y: vector[_common_fields]
24 │ │         };
   │ ╰─────────^
```

### PR DESCRIPTION
**Pull request by wrwg**
## Description

Fixes #13488
Fixes #14227

The original problem here is that the ability processor did not generate `copy` instructions for arguments of operators (like `a + b`).

The solution is a bit of a revision how the copy generation works, resulting in correct and potentially smaller and better to optimize code:

- The file format generator already generates copy based on live-var info. The only thing it cannot do is knowing whether a local need to be copied because it is borrowed.
- The ability processor continues to perform all copy checks, but only _generates_ a copy for those locals which are borrowed. The set of such locals is now extended by arguments of operators.

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?

Renamed the existing test case bug_13488 to bug_13488_missing_copy_for_add

### CODE REVIEW

**Comment by wrwg** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14292?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#14292** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14292?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#14282** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14282?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#14280** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14280?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#14276** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14276?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @wrwg and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by rahxephon89 — COMMENTED** [review]
In this closed PR, I added test cases `third_party/move/move-compiler-v2/tests/reference-safety/bug_13488_err.move` and `third_party/move/move-compiler-v2/transactional-tests/tests/no-optimize-only/bug_13488_v1_err.move` for other operations such as write_ref, vector operation and function call: https://github.com/aptos-labs/aptos-core/pull/13518/files#diff-4fef58e6ac3a394d71f1f766d2c4ce5502ad4f6857c1596366a07471dd0426c6. Could you add them to this PR as well?

**Comment by wrwg** [comment]
Thanks @rahxephon89 for pointing to PR #13518, I have integrated all the tests. With us aligning completely with v1 semantics, your PR starts making more sense. The current solution is similar but generates less copies in the IR, whereas the one in the older PR creates more.

**Review by rahxephon89 — APPROVED** [review]

**Comment by wrwg** [comment]
@vineethk it seems this PR regresses some of your coalescing tests, as there are no more locals to coalesce. In ordere to compensate, I tried explicit `copy x`, but not clear whether this keeps the tests working as intended.

File: third_party/move/move-compiler-v2/src/pipeline/reference_safety/reference_safety_processor_v3.rs [hunk 1]
```diff
@@ -1169,17 +1169,22 @@ impl LifetimeInfo for LifetimeState {
 // ===============================================================================
 // Display
 
-struct LabelDisplay<'a>(&'a FunctionTarget<'a>, &'a Label);
+struct LabelDisplay<'a>(&'a FunctionTarget<'a>, &'a Label, bool);
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/src/pipeline/reference_safety/reference_safety_processor_v3.rs:1172]: Document what the `bool` stands for.

File: third_party/move/move-compiler-v2/tests/variable-coalescing/consume_5.move [hunk 2]
```diff
@@ -22,9 +22,9 @@ module 0xc0ffee::m {
     public fun test_struct(p: bool, a: W) {
         let b = copy a;
         if (p) {
-            consume_(b);
+            consume_(copy b);
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tests/variable-coalescing/consume_5.move:25]: Not sure why we need this change. Could we please revert it to the original?

If this change is indeed useful, we could add another test with the changes.

**Review by vineethk — APPROVED** [review]
Approving assuming the comments are addressed.

File: third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs [hunk 3]
```diff
@@ -9,19 +9,18 @@
 //! - It infers the `AssignKind` in the assign statement. This will be `Move` if
 //!   the source is not used after the assignment and is not borrowed. It will
 //!   be Copy otherwise.
-//! - It inserts a `Copy` assignment for every function argument which is used later or borrowed
-//!   (same condition as above)
+//! - It inserts a `Copy` assignment for every function or operator argument which is not
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs:12]: Yes, misplace 'not'.

File: third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs [hunk 4]
```diff
@@ -9,19 +9,18 @@
...
 //! - It inserts a `Drop` instruction for values which go out of scope and are not
 //!   consumed by any call and no longer borrowed.
 //!
 //! For the checking part, consider the transformation to have happened,
 //! then:
 //!
-//! - Every copied value must have the `copy` ability
+//! - Every copied value must have the `copy` ability. This includes values for which we do
+//!   not generate a `Copy` instruction, but generally every temporary which is used later
+//!   or is borrowed.
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs:22]: Yes only the later. The Copy will still be inserted but later in the file format generator. Here we check whether it is allowed., and only insert it if the need for it is not derivable from simple liveness info.

[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs:22]: We still have to work out whether it is helpful besides file format generator and more naturally maintaining the stack order in the IR, but the new invariant subsequent phases can assume is: live_var info fully determines if a copy or move is needed when passing a parameter.

**Pull request merged**